### PR TITLE
Simplify CONTRIBUTING.rst

### PR DESCRIPTION
**Pull request by paolaperaza**
Adding a PR here that seeks to simplify Airflow's [CONTRIBUTING.rst](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst).

Specifically, this PR:

- Removes most sections under "Contributions" to a single "Welcome" section
- Simplifies the overview "Contribution Workflow" to actually match the steps provided
- Notably, this PR creates a new top-level `contributing` folder on apache/airflow, that then hosts 3 files: 
    - `CONTRIBUTE-CODE.rst` (converted from `contributing.rst`)
    - `CONTRIBUTE-DOCS.rst` (_new_)
    - `CONTRIBUTE-PROVIDERS.rst` (_new_)
    - `CONTRIBUTE-QUICK_START.rst` (renamed from `CONTRIBUTORS_QUICK_START.rst`
- `contributing.rst` will remain as a top-level directory, and act as a short resource that links out to the above docs

The idea here is that if a person wants to contribute to Airflow, we should give them tailored instructions depending on the part of the project they want to make changes to. Additionally, most of the actions I've taken here were talked about at our bi-weekly "Issue Triage" meeting ([Notes](https://docs.google.com/document/d/19y8AfZ56mvIo1AjxoFz0Ebzj_9VQJ6YTr0Acj64W1L4/edit)).

**Additional Notes:**
- The `CONTRIBUTE-DOCS.rst` file does not have much content at all - for now I didn't create new content, but just ported over existing language in what was `contributing.rst`. Would we be comfortable merging this PR as-is? The rest of the files should generally be looked at in more detail as well.
- I imagine these files are linked in quite a few places...given the new file paths I'm suggesting, we'll need to make sure to either create redirects or convert all existing references to the new paths.
- Need to re-link [this section in the README](https://github.com/apache/airflow#who-maintains-apache-airflow):
     > If you would like to become a maintainer, please review the [Apache Airflow committer requirements](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#guidelines-to-become-an-airflow-committer).

cc: @kaxil @vikramkoka @eladkal Before I move forward with this and polish what I have here, does this structure I'm proposing make sense/is feasible? I'm not entirely familiar with the typical structure of these .rst files, so I want to make sure this is "ok" to do in a single PR. Let me know on general thoughts, and I'll do some more work on this.

### CODE REVIEW

**Review by kaxil — COMMENTED** [review]
Is this ready for review @paolaperaza ?

**Comment by marcosmarxm** [comment]
Amazing work @paolaperaza I started to work on a how-to guide to create a new provider in #15061 (structure files, examples of initial files etc). Initially, I placed it inside the `docs/apache-airflow-providers` with other providers' documentation. Probably with this new document will be better placed together.

**Comment by uranusjr** [comment]
Would it be worthwhile to still have a `CONTRIBUTING` file? I think GitHub detects this particular file name and displays it in new PRs. https://docs.github.com/en/github/building-a-strong-community/setting-guidelines-for-repository-contributors#adding-a-contributing-file

**Comment by potiuk** [comment]
> Would it be worthwhile to still have a `CONTRIBUTING` file? I think [GitHub detects this particular file name and displays it in new PRs](https://docs.github.com/en/github/building-a-strong-community/setting-guidelines-for-repository-contributors).

Very good point. We should have and it should be a directory to all the split documentation.

It's also part of the "commity" checklist https://github.com/apache/airflow/community

**Comment by paolaperaza** [comment]
@potiuk @uranusjr Yes! I agree - @ashb actually suggested that at our Issue Triage meeting last week and we all agreed it was a good idea. To confirm, this means that I will:

- Keep `CONTRIBUTING.rst`, which will act as a directory to the rest of "Contributing" content (i.e. the other 3 files referenced in this PR).
- Those files will be in a separate `contributing` directory (top-level).

This may be in a separate PR, but I'd also like to get the directory of that content (equivelent to content in CONTRIBUTING.rst) on https://airflow.apache.org/docs/, since our docs site is where most people in the community go for information. Ultimately looking to do something like [this from Ansible](https://docs.ansible.com/ansible/latest/dev_guide/index.html), but a bit more consolidated.

---

@marcosmarxm Thanks for the note! I'll comment in that PR, but I think it's fair to merge https://github.com/apache/airflow/pull/15061 before this PR, and I'll rebase and incorporate those changes here afterwards, if you all agree.

**Comment by potiuk** [comment]
Cool!

**Comment by paolaperaza** [comment]
@kaxil @potiuk @vikramkoka A bit delayed here, but I THINK this PR is finally ready for review 😊   My latest changes include:

- Pulling in most recent commits to `contributing.rst` (since I changed the file name, I couldn't rebase and merge).
- Wrote new content for `contribute-docs.rst`

**Question:**
Did we want to pull in @marcosmarxm changes to my new `contribute-providers.rst` file, or keep that file under the `apache-airflow-providers` directory? His currently lives here: https://github.com/apache/airflow/blob/932f8c2e9360de6371031d4d71df00867a2776e6/docs/apache-airflow-providers/howto/create-update-providers.rst

The `contribute-providers.rst` file that I created was really just extracting provider-related info from the original `contributing.rst`. I could also just delete it?

Let me know what you two think so we can keep the ball moving. Appreciate it!

**Comment by potiuk** [comment]
Hey @paolaperaza - can you please rebase this one ? I'd love to get this one merged and will have more time to review those kind of changes now :)

**Comment by potiuk** [comment]
> Did we want to pull in @marcosmarxm changes to my new `contribute-providers.rst` file, or keep that file under the `apache-airflow-providers` directory? His currently lives here: https://github.com/apache/airflow/blob/932f8c2e9360de6371031d4d71df00867a2776e6/docs/apache-airflow-providers/howto/create-update-providers.rst

I think let's merge yours first and the we can as @marcosmarxm  to rebase his changes n top.

File: contribute/CONTRIBUTE-DOCS.rst [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+What Should I Contribute?
+----------------------
+Do you want to contribute to Airflow docs, but you're not sure exactly where? Consider any of the following:
+
+- Typo fixes
+- Additional information about existing functionality
+- Documentation about a new feature
+- Images or screenshots that help provide users context
+- New tutorials
+
```
Comment (vikramkoka) [inline contribute/CONTRIBUTE-DOCS.rst:42]: For ideas, you can browse the open identified issues:

- `All GitHub Documentation Issues  <https://github.com/apache/airflow/issues?q=is%3Aopen+is%3Aissue+label%3Akind%3Adocumentation>`__

File: contribute/CONTRIBUTE-CODE.rst [hunk 2]
```diff
@@ -0,0 +1,934 @@
...
+follow `Contributors Quick Start <https://github.com/apache/airflow/blob/master/contribute/CONTRIBUTE_QUICK_START.rst>`__ first for easy, step-by-step guidelines on how to set up your development
+environment and make your first contribution.
+
+If you're interested in contributing to other parts of the project, refer to
+`Contribute to Airflow Documentation <https://github.com/apache/airflow/blob/master/contribute/CONTRIBUTE-DOCS.rst>`__ or `Contribute to Airflow Providers <https://airflow.apache.org/docs/apache-airflow-providers/index.html#creating-your-own-providers>`__.
+
+Where to Contribute
+--------------------
+To get started, decide where or what you want to contribute. For ideas, you can browse:
+
```
Comment (vikramkoka) [inline contribute/CONTRIBUTE-CODE.rst:36]: - `Good First Issues <https://github.com/apache/airflow/issues?q=is%3Aopen+is%3Aissue+label%3A%22good%20first%20issue%22>`__

**Review by vikramkoka — APPROVED** [review]
@paolaperaza This looks great, just left a couple of comments to help early contributors find issues that they could help with.

**Review by potiuk — APPROVED** [review]
Agree. Great job @paolaperaza

File: contribute/CONTRIBUTE-DOCS.rst (hunk 1 above)
Comment (potiuk) [reply contribute/CONTRIBUTE-DOCS.rst:42]: ~Actually we can do a bit better https://github.com/apache/airflow/contribute - is a dedicated GitHub link for that label + it has `CONTRIBUTING.rst" link + nice invite explaining what those issues are.~

Comment (potiuk) [reply contribute/CONTRIBUTE-DOCS.rst:42]: Ah no.. This is also for `kind:doc` - withdraw it.

File: contribute/CONTRIBUTE-PROVIDERS.rst [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+There are a number of extras that can be specified when installing Airflow. Those
+extras can be specified after the usual pip install - for example
+``pip install -e .[ssh]``. For development purpose there is a ``devel`` extra that
+installs all development dependencies. There is also ``devel_ci`` that installs
+all dependencies needed in the CI environment.
+
+This is the full list of those extras:
+
+  .. START EXTRAS HERE
```
Comment (potiuk) [inline contribute/CONTRIBUTE-PROVIDERS.rst:55]: We also need to make one CODE change here.  this document gets automatically updated whenever extras are added, so you also need to change name of the file here:

https://github.com/apache/airflow/blob/master/scripts/ci/pre_commit/pre_commit_insert_extras.py#L62

From `CONTRIBUTING.rst` to `CONTRIBUTE-PROVIDERS.rst`

**Review by potiuk — CHANGES_REQUESTED** [review]
One small change about script automation needed.

File: contribute/CONTRIBUTE-DOCS.rst (hunk 1 above)
Comment (paolaperaza) [reply contribute/CONTRIBUTE-DOCS.rst:42]: Good call to add this, @vikramkoka ! Made this change in my latest commit.

File: contribute/CONTRIBUTE-CODE.rst (hunk 2 above)
Comment (paolaperaza) [reply contribute/CONTRIBUTE-CODE.rst:36]: Good call here too - added

File: contribute/CONTRIBUTE-PROVIDERS.rst (hunk 3 above)
Comment (paolaperaza) [reply contribute/CONTRIBUTE-PROVIDERS.rst:55]: Oh good call @potiuk - wouldn't have seen that. Done!

**Comment by paolaperaza** [comment]
@vikramkoka @potiuk Thank you both for reviewing this! I just made both of the changes you all suggested. One last question: I did another global search on `contributing.rst` and found a lot of references to that file with a specific section. Do you all think I need to adjust those, or is it find that we're still linking out to `contributing.rst` and people can find the info they need?

For example, our README states:

> You can read more about the naming conventions used
> in [Naming conventions for provider packages](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#naming-conventions-for-provider-packages)

Should I look through all of these and update to `CONTRIBUTE-PROVIDERS.rst` or `CONTRIBUTE-CODE.rst` etc. where appropriate?

**Comment by potiuk** [comment]
> Should I look through all of these and update to `CONTRIBUTE-PROVIDERS.rst` or `CONTRIBUTE-CODE.rst` etc. where appropriate? I did make that type of change in https://github.com/apache/airflow/blob/master/BREEZE.rst.

Absolutely. Good spot! Currently we have a major GitHub Actions outage, I hope it's going to be solved soon, but happy to review it approve and merge soon after it's finished and you update the references! https://www.githubstatus.com/incidents/zbpwygxwb3gw

**Comment by jhtimmins** [comment]
@paolaperaza can you resolve these conflicts? CI is generally stable again, so we should be able to get this merged.

@potiuk can you validate that your requested changes have been addressed?

**Comment by potiuk** [comment]
Yep. Will do it after @paolaperaza rebases it.

**Pull request closed without merging**